packages/protocol/lib/layout.ts

```diff
@@ -0,0 +1,216 @@
+
+const web3 = new Web3(null)
+
+// Inlined from OpenZeppelin SDK since its not exported.
+interface Artifact {
+  abi: any[]
+  ast: any
+  bytecode: string
+  compiler: any
+  contractName: string
+  deployedBytecode: string
+  deployedSourceMap: string
+  fileName: string
+  legacyAST?: any
+  networks: any
+  schemaVersion: string
+  source: string
+  sourceMap: string
+  sourcePath: string
+  updatedAt: string
+}
+
+// Inlined from OpenZeppelin SDK since its not exported.
+interface TypeInfo {
+  id: string;
+  kind: string;
+  label: string;
+  valueType?: string;
+  length?: number;
+  members?: StorageInfo[];
+  src?: any;
+}
+
+// Inlined from OpenZeppelin SDK since its not exported.
+interface StorageInfo {
+  label: string;
+  astId: number;
+  type: any;
+  src: string;
+  path?: string;
+  contract?: string;
+}
+
+// getStorageLayout needs an oz-sdk Contract class instance. This class is a
+// subclass of Contract from web3-eth-contract, with an added .schema member and
+// several methods.
+//
+// Couldn't find an easy way of getting one just from contract artifacts. But
+// for getStorageLayout we really only need .schema.ast and .schema.contractName.
+const addSchemaForLayoutChecking = (web3Contract: Web3Contract, artifact: any): ZContract => {
+  // @ts-ignore
+  const contract = web3Contract as Contract
+  // @ts-ignore
+  contract.schema = {}
+  contract.schema.ast = artifact.ast
+  contract.schema.contractName = artifact.contractName
+  return contract
+}
+
+const makeZContract = (artifact: any): ZContract => {
+  const contract = new web3.eth.Contract(artifact.abi)
+
+  return addSchemaForLayoutChecking(contract, artifact)
+}
+
+export const getLayout = (artifact: Artifact, artifacts: BuildArtifacts) => {
+  const contract = makeZContract(artifact)
+
+  return getStorageLayout(contract, artifacts)
+}
+
+const selectIncompatibleOperations = (diff: Operation[]) =>
+  diff.filter(operation => operation.action !== 'append')
+
+export interface CompatibilityInfo {
+  contract: string
+  compatible: boolean
+  errors: string[]
+}
+
+const operationToDescription = (operation: Operation) => {
+  let message: string
+
+  const updated = operation.updated
+  const original = operation.original
+
+  switch (operation.action) {
+    case 'typechange':
+      message = `variable ${updated.label} had type ${original.type}, now has type ${updated.type}`
+      break
+    case 'insert':
+      message = `variable ${updated.label} was inserted`
+      break
+    case 'pop':
+      message = `variable ${original.label} was removed`
+      break
+    case 'delete':
+      message = `variable ${original.label} was removed`
+      break
+    case 'rename':
+      message = `variable ${updated.label} was renamed from ${original.label}`
+      break
+    case 'append':
+      message = `variable ${updated.label} was appended`
+
+    default:
+      message = `unknown operation ${operation.action}`
```

Comment:
Since it's an unknown operation, it's not clear which, if any label is defined. This shouldn't happen anyways, but one could imagine an unkown operation being "appendy" where the original label is undefined, or "deletey" where the updated label is undefined.